### ISSUE TITLE
V3.6.1 huge performance drop

### ISSUE THREAD

**Issue by mirekratman** [bcosca/fatfree#1030]
Hi
Today I've updated FF to v.3.6.1 and immediately realized a huge perrformance drop in website loading. Didn't check deeply the problem yet but simply changing only framework version from 3.5.1 to 3.6.1 the loading time increased from about 120-150ms up to 5000-5500 ms. Checked it few times to avoid some mistake but always had same result.

It can be something with config files loading because when i removed all configs loading time went down almost to the same level like in v3.5.1. Version 3.6.0 seems work fine like 3.5.1.

Will try to share more info after more investigation.

**Comment by ikkez** [comment]
Can you try once more with the latest dev-version on https://github.com/bcosca/fatfree-core? Because there were some little fixes included that might have caused the issue.

**Comment by mirekratman** [comment]
Same situation - loading time around 5500 ms. Switching back to 3.5.1 or 3.6.0 solving the problem.
Looks like there can be some issue related to base.php->config(). measured loading time of each operation and i gave me huge difference.

**Comment by ikkez** [comment]
Do you use the `$allow` parameter on ->config?

**Comment by mirekratman** [comment]
Yes, its used

**Comment by ikkez** [comment]
I'm not entirely sure what's happening, but I don't see the issue on this and have not noticed any performance issues on my site yet. Nevertheless I think what could be optimized is this part:

https://github.com/bcosca/fatfree-core/blob/master/base.php#L1863-L1868
```php
if ($allow) {
	$match['lval']=Preview::instance()->
		resolve($match['lval']);
	$match['rval']=Preview::instance()->
		resolve($match['rval']);
}
```
Instead of resolving each left and right config entry part on its own, we can parse the whole file at once at this point https://github.com/bcosca/fatfree-core/blob/master/base.php#L1846 and render it as template.

```php
preg_match_all(
'/(?<=^|\n)(?:'.
	'\[(?<section>.+?)\]|'.
	'(?<lval>[^\h\r\n;].*?)\h*=\h*'.
	'(?<rval>(?:\\\\\h*\r?\n|.+?)*)'.
')(?=\r?\n|$)/',
$this->read($file),
$matches,PREG_SET_ORDER);
```

becomes

```
$template=Preview::instance();
preg_match_all(
'/(?<=^|\n)(?:'.
	'\[(?<section>.+?)\]|'.
	'(?<lval>[^\h\r\n;].*?)\h*=\h*'.
	'(?<rval>(?:\\\\\h*\r?\n|.+?)*)'.
')(?=\r?\n|$)/',
$allow
	? $template->resolve($template->parse($file))
	: $this->read($file),
$matches,PREG_SET_ORDER);
```
It's just an idea.
I'm not sure if that might raise other issues, but maybe worth a try.

**Comment by bcosca** [comment]
My concern with this approach is that an .ini variable defined in one line cannot use a setting defined in a previous line:

a=123
b={{ @a }}

**Comment by ikkez** [comment]
good point. I didn't think of that.

**Comment by stehlo** [comment]
May I ask, what is the real point in using this overcomplicated syntactic sugar within the config ini files, when it is so easy to do this directly in php files?

Every time some poorly tested new feature or change is introduced, there is a penalty to pay. I am really scared of this trend.

For around six months I have been on a non-php project and as I am returning to F3-based ones now, I indeed wonder whereto we got in between...

**Comment by ikkez** [comment]
Actually, this [code](https://github.com/bcosca/fatfree-core/commit/5707abfbd8e27b13d719beed3517a6ef718a58c8#diff-22b1ffa797ed0997a20765c7cf8a6549R1571) is part of the framework since 3.5 (more than 2 years ago). I'm just looking for ways to improve things, but I understand your concerns.

**Comment by stehlo** [comment]
@ikkez 

> but I understand your concerns

I hope you do [and not just say].

**Comment by stehlo** [comment]
Also, @ikkez ,

> more than 2 years ago

The age of code doesn't reflect its potential grave problems, as in the following case: https://github.com/bcosca/fatfree-core/issues/214

To the contrary, untested changes of mediocre qualities can lay very deeply unnoticed – until one discovers that previously working functionality is utterly broken. That's very bad.

The issue is exacerbated by the fact that these code changes are arriving in big chunks and hit unsuspecting recipients with an unknown and possibly very wide scope.

At this point I am questioning, whether Bong Cosca @bcosca didn't make a mistake by giving up his previous mode of operation, i.e. not allowing other people to change his code [of FatFree].

I am **really, really seriously worried**, because F3 is potentially becoming unreliable as all open-source open-development products – whereby "open-source" is not a problem, but "open-development" yes and a big one at that.

**Comment by ikkez** [comment]
Hey @stehlo I can read that you are very excited, not just because you have found a serious bug, no one noticed before yet. Of course that is bad and should not happen, even when refactoring parts of the framework. But the complexity isn't always visible at the first sight and believe me when I say that I am really really into keeping things running stable and fast than adding new features. I literally inspect very single commit made by xfra35 and bong twice, since more than 3 years now and I would say that I've fixed more than I've broken or added here - and since I do run business on this codebase, I'm really interested in finding issues in code changes more than you might think. SO finally I believe it became even more stable and reliable than ever before. There're some bugs for sure - very software has them, but I think our very little core team is able to spot those faster than a single person (you're still the best @bcosca 😉 )

**Comment by stehlo** [comment]
Hello @ikkez 

> you are very excited

Most definitely not "excited"!

It is more of a mixture of furiousness and worries, because the update screwed up a very important enterprise project of a large client of mine – well, at least, until it has been fixed.

**Comment by ikkez** [comment]
If you need support with at, join us on [slack](https://fatfreeframework-slack.herokuapp.com/) and maybe have a look some [important changes](http://slides.com/ikkez/fat-free-v3-6#/2). We'll have a look at bcosca/fatfree-core#214 in the meantime.

**Comment by stehlo** [comment]
I am specifically refraining from involvement in chat engines, as this would go against the very notion of what I just wrote above. I prefer to donate/pay money to/for quality products and communicate in private. (These comms here are a glaring rare exceptions – few and far between.)

**Comment by ikkez** [comment]
>Things like this: bcosca/fatfree-core@99e1762 are exactly what I mean. A production-ready code shouldn't become a playground for testing half-arsed ideas.

I totally agree with you, and I whether like that commit nor the fact that the master branch is used as test platform. I would probably work with more branches, release more and smaller updates and use a different workflow, but I'm following the author's (bong) project habits here.

I just meant slack as option for private communication, but do as you like.

**Comment by SylwesterZarebski** [comment]
I'm not sure, if it is the same issue, but after upgrading to 3.6.2 version (from 3.6-dev), i'm observing massive slowness of every page refresh even non existant ones (404).
On development hosting with Atom CPU every page opening is lasting more than 10 seconds.

With previous version opening non existant files lasts about 300ms.

**Comment by SylwesterZarebski** [comment]
Quick analysis bring a culprit which is `resolve()` method. 3.6.0 version is way simpler and quicker, at least one order of magnitude.

**Comment by ikkez** [comment]
Hi again. Finally I also noticed massive slowdowns under some circumstances and I investigated the issue a little more. You're absolutely right @SylwesterZarebski, the resolve method  [@ 3.6.0](https://github.com/bcosca/fatfree-core/blob/v3.6.0/base.php#L2787-L2794) was indeed much shorter, but it was adjusted to also be able to parse and resolve much bigger strings like whole templates served from a database etc. However, the portion of escaping the injected hive was new and should be disabled at all for "just" parsing config entries. I've fixed this in https://github.com/bcosca/fatfree-core/commit/efa1ea759f235268698a8ca82a9d3a9c49b01fb5 now.
There's also another issue going on that could araise when using config includes like this:

```ini
[configs]
app/routes.ini = FALSE
app/page_defaults.ini = FALSE
```

The `configs` and other sections are treated as "command" entries, not just simple key-value pairs that are pushed into the hive. These lacked on support for proper variable casting, which lead to a method parameter of `(string) "FALSE"` instead of `(bool) FALSE` - and this finally screwed up further processing and turned dynamic configs ON even if you set it to false. I've fixed this in https://github.com/bcosca/fatfree-core/commit/28e57856480f89728312a8ae93c086cbb9595857
Both fixes now reduces my config parsing time from 9000ms back to 20ms,.. and with dynamic config turned on it was still a nice 60ms (~330 config entries for testing).
